### PR TITLE
Java api fixes

### PR DESCRIPTION
**Pull request by lcc**
This patch fixes the following misuses of the java api:

1. Meaningless Close:  In several classes, close(), specified in Closeable interface, has no effect and other methods in those classes can be called after close() without IOException.

2. Flushes OutputStream before using the underlying ByteArrayOutputStream: when an OutputStream (or its subclass) instance is built on top of an underlying ByteArrayOutputStream instance, it should be flushed or closed before the underlying instance's toByteArray() is invoked. Failing to fulfill this requirement may cause toByteArray() to return incomplete contents.

3. Unsynchronized addAll:  The behavior of Collection.addAll() is undefined if the specified source collection is modified while the operation is in progress.

Fixes regarding (1) only have a consequence in coding-style/clarity, but IMO the impact on clarity is quite big as there are some implementations of closeQuietly methods that have no effect (since the close method, for these classes, never throw IOException).

(2) and (3) may lead to incoherent behaviour.

### CODE REVIEW

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java [hunk 1]
```diff
@@ -473,10 +474,7 @@ public static int getBufferSize(String contentLength) {
 	 * @param is The input stream to close.
 	 */
 	public static void closeQuietly(InputStream is) {
-		try {
-			if (is != null)
-				is.close();
-		} catch (IOException e) {}
+
```
Comment (jamesbognar) [inline juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java:477]: This code shouldn't be removed.  The IOUtils.closeQuietly(...) methods are general utility methods for closing streams.

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java [hunk 2]
```diff
@@ -578,8 +570,6 @@ public static void close(Object...o) throws IOException {
 		IOException ex = null;
 		for (Object o2 : o) {
 			try {
-				if (o2 instanceof InputStream)
-					((InputStream)o2).close();
```
Comment (jamesbognar) [inline juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java:582]: The IOUtils.close(Object...) method is also a general utility method that's supposed to work on InputStreams.  So this code also should not be removed.

File: juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java [hunk 3]
```diff
@@ -55,7 +55,6 @@ public final void flush() throws IOException {
 
 	@Override /* OutputStream */
 	public final void close() throws IOException {
-		os.close();
```
Comment (jamesbognar) [inline juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java:58]: Why was this removed?

**Comment by jamesbognar** [comment]
Changes look good except for some code that shouldn't have been removed.

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java (hunk 1 above)
Comment (lcc) [reply juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java:477]: This method calls is.close() and is is an InputStream. Java's documentation states that [InputStream close()](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#close--) doesn't do anything.

Since is.close() doesn't do anything (including throwing an IOException), having a closeQuietly method is IMHO a little confusing.

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java (hunk 2 above)
Comment (lcc) [reply juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java:582]: As stated above, there doesn't seem to be any reason to call close on an InputStream.

File: juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java (hunk 3 above)
Comment (lcc) [reply juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java:58]: os is an OutputStream which has the same behaviour for the close method as the InputStream, as stated [here](https://docs.oracle.com/javase/8/docs/api/java/io/OutputStream.html#close--).

**Review by lcc — COMMENTED** [review]
Thanks for the prompt response James, if anything, let me know.

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java (hunk 1 above)
Comment (jamesbognar) [reply juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java:477]: The Java documentation is misleading.  The close() method on the InputStream class itself may be a no-op, but it is very much not a no-op on subclasses of InputStream (FileInputStream for example).

File: juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java (hunk 3 above)
Comment (jamesbognar) [reply juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java:58]: Same reason as above.  The close() method on OutputStream may be a no-op, but it is not a no-op on subclasses.

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java (hunk 1 above)
Comment (lcc) [reply juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java]: Done.

File: juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java (hunk 2 above)
Comment (lcc) [reply juneau-core/juneau-marshall/src/main/java/org/apache/juneau/internal/IOUtils.java]: Done.

File: juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java (hunk 3 above)
Comment (lcc) [reply juneau-rest/juneau-rest-server/src/main/java/org/apache/juneau/rest/util/FinishableServletOutputStream.java]: Done.

**Comment by jamesbognar** [comment]
Pull request has been merged.  Sorry for the delay. 

Thanks for your contribution!
